modules/augain/augain.c

```diff
@@ -0,0 +1,165 @@
+/**
+ * @file augain.c  Audio gain
+ *
+ * Copyright (C) 2024 Juha Heinanen
+ */
+
+#include <stdlib.h>
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+
+
+/**
+ * Audio gain module
+ *
+ * This module can be used to increase volume of audio source,
+ * for example, microphone.
+ *
+ * Sample config:
+ *
+ \verbatim
+  augain            1.5
+ \endverbatim
+ */
+
+
+struct augain_enc {
+	struct aufilt_enc_st af;  /* base class */
+};
+
+
+static double gain = 1.0;
+
+
+static int conf_get_float(const struct conf *conf, const char *name,
+			  double *val)
+{
+	struct pl opt;
+	int err;
+
+	if (!conf || !name || !val)
+		return EINVAL;
+
+	err = conf_get(conf, name, &opt);
+	if (err)
+		return err;
+
+	*val = pl_float(&opt);
+
+	return 0;
+}
+
+
+static void enc_destructor(void *arg)
+{
+	struct augain_enc *st = arg;
+
+	list_unlink(&st->af.le);
+}
+
+
+static int encode_update(struct aufilt_enc_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct augain_enc *st;
+	(void)ctx;
+	(void)af;
+	(void)au;
+
+	if (!stp || !prm)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), enc_destructor);
+	if (!st)
+		return EINVAL;
+
+	conf_get_float(conf_cur(), "augain", &gain);
+
+	*stp = (struct aufilt_enc_st *)st;
+
+	return 0;
+}
+
+
+static int encode_frame(struct aufilt_enc_st *st, struct auframe *af)
+{
+	size_t i;
+	unsigned int abs_sample, highest_abs_sample = 0;
+	double encode_gain = gain;
+	double highest_possible_gain;
+	int16_t sample, gained_sample;
+
+	if (!st || !af || !af->sampv || !af->sampc)
+		return EINVAL;
+
+	if (af->fmt != AUFMT_S16LE) {
```

Comment:
This check should be moved to `encode_update` (`prm->fmt`), not needed to execute for every frame.

Reply:
I moved `af->fmt` check but was not sure if also `if (!st || !af || !af->sampv || !af->sampc)` could have been moved.